### DISCUSSION TITLE
Meson build system

### DISCUSSION THREAD

**Discussion by HDembinski** [impy-project/chromo#107] [Ideas]
I researched Meson in the last days and want to write down some notes.

@afedynitch likes meson. I also only hear good things about it, although I don't have first-hand experience yet. When we made the port from the old make-based build to cmake, I was a strong advocate of cmake, since it is a mature tool that I understand well enough so that I can write simple functions and macros. As expected, that turned out to be useful, since we have to do some shenanigans to make the code generation via f2py work.

However, now that we have a working baseline in CMake, we can give meson a try. While cmake and meson are not exactly similar, they are based on the same high-level concepts so that porting cmake to meson is easier than porting the original make system. If meson does not work for us, we also still have a baseline that works, so the meson build can be developed as a branch in parallel without hindering development on the main branch.

### Why is moving to meson useful at all at this point? 

Let's not talk about syntax, I am not a fan of the SCREAMCASE employed by cmake, but overall, I don't think the cmake syntax is bad. It is fairly verbose which is a good for readability, and documentation is good. So let's focus on things that meson provides but cmake does not.

#### Meson is part of the Python ecosphere

[Meson](https://github.com/mesonbuild/meson) is a Python application and distributed via PyPI. This makes installing the latest and greatest meson version easy and it makes it easy to write extensions that are build with meson. We currently use setuptools to build our project with a hook to let cmake do the actual work. This works but is a bit clunky. If we use meson, we can potentially switch to [meson-python](https://github.com/mesonbuild/meson-python), a real PEP 517 compliant build backend, which should work out-of-the-box.

#### Meson can download dependencies for us

Meson can be configured to download and install dependencies automatically, like a packet manager.
https://mesonbuild.com/Wrap-dependency-system-manual.html
This means we may get rid of some git submodules. Projects need to be wrapped for automatic inclusion, and the code to do this can be uploaded to a shared DB. That's pretty nice.
https://mesonbuild.com/Wrapdb-projects.html

#### Meson seems mature enough now

Meson just reached v1.0.0, and its list of users contains some big players.
https://mesonbuild.com/Users.html
Lots of GNOME projects use it, and very important for us, Scipy is currently making the switch. Scipy is a lighthouse for us, because they face the same difficulties as us; they need to compile a mix of Fortran, C, C++ into a Python package and distribute it. 

It's documentation is fairly extensive and comprehensible. It seems to have all the feature we need, also for code generation https://mesonbuild.com/Generating-sources.html

**Comment by HDembinski** [comment]
I have an experimental meson port of pyhepmc here: https://github.com/scikit-hep/pyhepmc/pull/57

There are still some issues to work out before we can use meson everywhere. The main thing is that meson-python (the build backend used by pip) does not support editable builds yet.

**Reply by eli-schwartz** [reply]
Editable installs support was merged into meson-python but you currently need to install from git until a new release is cut.

Testing is welcomed. :)

**Reply by HDembinski** [reply]
Hi @eli-schwartz, I actually tried the editable feature in the trunk and it did not work for my case, unfortunately. I will post an issue about that.

**Reply by eli-schwartz** [reply]
Cool, thanks for testing. :)
